### PR TITLE
Refactor ProcessNewBlock to reduce code duplication

### PR DESCRIPTION
**Pull request by rebroad**
There are probably a few issues with this code (maybe there's even a reason this code is duplicated as it currently is), so apologies in advance that I'm still a little (maybe very) bad with C++

### CODE REVIEW

File: src/net_processing.cpp [hunk 1]
```diff
@@ -2309,6 +2311,19 @@ void PeerManagerImpl::ProcessGetCFCheckPt(CNode& peer, CDataStream& vRecv)
     m_connman.PushMessage(&peer, std::move(msg));
 }
 
+void PeerManagerImpl::ProcessBlock(CNode& pfrom, const CChainParams& chainparams, const std::shared_ptr<const CBlock> pblock,
```
Comment (sipa) [inline src/net_processing.cpp:2314]: Nit: you can turn `pblock` into a const reference argument (`const std::shared_ptr<const CBlock>& pblock`), which will avoid an atomic shared_ptr copy at runtime.

**Comment by sipa** [comment]
Looks obviously correct. utACK 89e5e5a67977588218c3c4309d9291b0cc5dd4ad

File: src/net_processing.cpp (hunk 1 above)
Comment (practicalswift) [inline src/net_processing.cpp:2314]: `chainparams` is not being used in the function body. `m_chainparams` is used however.

Comment (rebroad) [reply src/net_processing.cpp:2314]: curious... I wonder why the tests aren't failing then...

Comment (MarcoFalke) [reply src/net_processing.cpp:2314]: >  I wonder why the tests aren't failing then

We have no tests to detect unused parameters, IIRC

Comment (rebroad) [reply src/net_processing.cpp]: Thanks. changed. If I may ask, why doesn't ProcessNewBlock() also do this?

Comment (rebroad) [reply src/net_processing.cpp]: I thought the compiler would at least give a warning - I've seen it warn about unused variables before, so I wonder why it doesn't do this for unused parameters too.

Comment (MarcoFalke) [reply src/net_processing.cpp]: Probably just a typo

File: src/net_processing.cpp [hunk 2]
```diff
@@ -3483,20 +3490,13 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
...
             // protections in the compact block handler -- see related comment
             // in compact block optimistic reconstruction handling.
-            m_chainman.ProcessNewBlock(m_chainparams, pblock, /*fForceProcessing=*/true, &fNewBlock);
-            if (fNewBlock) {
-                pfrom.nLastBlockTime = GetTime();
-            } else {
-                LOCK(cs_main);
-                mapBlockSource.erase(pblock->GetHash());
-            }
+            ProcessBlock(pfrom, pblock, /*fForceProcessing=*/true);
```
Comment (promag) [inline src/net_processing.cpp:3499]: Why not just `ProcessBlock(pfrom, pblock)`?

**Review by promag — COMMENTED** [review]
Code review ACK 9a0653553a0ec403b4e7c6713466e0c7fa10ec94.

**Review by theStack — APPROVED** [review]
Code-review ACK 9a0653553a0ec403b4e7c6713466e0c7fa10ec94 🌴

**Comment by MarcoFalke** [comment]
ACK 9a0653553a0ec403b4e7c6713466e0c7fa10ec94 💻

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 9a0653553a0ec403b4e7c6713466e0c7fa10ec94 💻
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiYnAv/Q8in67v6vQkKBao81ya/jLVKWU/9q7/X3Hpcvsqy4eD0zr+dr/sVG2Yw
YBDoA2Owy9M2Y0e+gEBOzQmYfwgsMeeotFaRrZfFdUCrxQ9LYcZlU93FlNcdJEFk
NlbQwUlmuQ8PQ2ZuDoJQetkGvROh0SscH4F755GDxP3PUoG4FoE3rkJdvIucQi6M
LNH1kAcOQua8ftWJsrIN/tmnP4NFj5MM+baqwbd4rpPUE967/jjAPF1mrwOFJa/c
orRnm65vi9MIQsq1zCJDv0WOxFlTnMZCcNy743/oClnklAyxByLxaNOkpSSb1tz1
nZxiLw3ApRkIRxuZpgJUdCfNmLFbI3wfoU0xzL3bP84AGnUocyVyc2iFPRZcPLW8
5OXzzUowvVCIRa4Nn3bv7bHbwzhEJjKa0scm8+7wzX8VQjC1EmwHCJ1uvbU3I4vJ
kNE2+XgSlNA9vEqzxVn6QiioiI/gfO8RT0svmrr+tcpBbVnTrqQ5aDUW1+KGNR2/
lVlWmxbz
=4Pl6
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `a6152b70c0c6ef60deb64a857a8b598107e67ea657a7702b710dc476876d1c76  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108a6152b70c0c6ef60deb64a857a8b598107e67ea657a7702b710dc476876d1c76f010c1df6b338d51ee4cafbc7f7b4c67900308fff010139898e98bf3cd6967be1291b0efecd508f0206ee8c9d7e5d61084bd0be40fe49b74bcef660a7ef27b545ddae4c5fafab48d6008f104607d2ab9f00820919d070bd90d560083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010a1365af5e36a3c13b2d212c1d5c44fd108f104607d2ab9f00868be8b7ba24c250e0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010d111dd4c81539fab6c792d55096719fd08f104607d1cdef008709366ecf0ef2d2e0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010fac7040e071c7206b2d96ab49e4b41fd08f120e99a81c6e6dc8771bb33af06bbc64d55f37f66aa1db16d29ee5c6228618b209a08f104607d2ab8f008efe870910fa97c7c0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

**Comment by fanquake** [comment]
Note that this PR changed the file permissions of `src/net_processsing.cpp` for no reason.

File: src/net_processing.cpp [hunk 3]
```diff
@@ -451,6 +451,8 @@ class PeerManagerImpl final : public PeerManager
 
     void ProcessGetData(CNode& pfrom, Peer& peer, const std::atomic<bool>& interruptMsgProc) EXCLUSIVE_LOCKS_REQUIRED(peer.m_getdata_requests_mutex) LOCKS_EXCLUDED(::cs_main);
 
+    void ProcessBlock(CNode& pfrom, const std::shared_ptr<const CBlock>& pblock, bool fForceProcessing);
```
Comment (jnewbery) [inline src/net_processing.cpp:454]: New functions should have doxygen comments.

File: src/net_processing.cpp [hunk 4]
```diff
@@ -2309,6 +2311,18 @@ void PeerManagerImpl::ProcessGetCFCheckPt(CNode& peer, CDataStream& vRecv)
     m_connman.PushMessage(&peer, std::move(msg));
 }
 
+void PeerManagerImpl::ProcessBlock(CNode& pfrom, const std::shared_ptr<const CBlock>& pblock, bool fForceProcessing)
+{
+    bool fNewBlock = false;
```
Comment (jnewbery) [inline src/net_processing.cpp:2316]:
```suggestion
    bool new_block{false};
```

**Review by jnewbery — COMMENTED** [review]
Functionality looks correct. It would have been nice to follow the style guide for the new code.

**Comment by arslan-raza-143** [comment]
Duplicate of # @MarcoFalke

**Pull request merged**